### PR TITLE
Reconcile standalone Stargate auth schema

### LINKED ISSUE #297: K8SSAND-1170 ⁃ e2e test flakes in Stargate tests
**What happened?**
We have been observing periodic e2e test failures lately. `CreateMultiStargateAndDatacenter` is one of the tests.

When deploying Stargate on its own, i.e., not part of a K8ssandraCluster, and with mulitple racks it can cause schema a ConfigurationException.

I investigated a recent failure. The test failed on a CassandraDatacenter readiness check:

```
Error:     suite_test.go:761: check that datacenter dc1 in cluster kind-k8ssandra-0 is ready
    suite_test.go:763: 
        	Error Trace:	suite_test.go:763
        	            				suite_test.go:530
        	            				suite_test.go:236
        	Error:      	Condition never satisfied
        	Test:       	TestOperator/CreateMultiStargateAndDatacenter
        	Messages:   	timed out waiting for datacenter dc1 to become ready
```

I downloaded the logs and saw this exception in a couple of the Cassadra pods' logs:

```
INFO  [MigrationStage:1] 2022-01-13 17:37:57,798 ColumnFamilyStore.java:432 - Initializing data_endpoint_auth.token
ERROR [MigrationStage:1] 2022-01-13 17:37:59,046 CassandraDaemon.java:244 - Exception in thread Thread[MigrationStage:1,5,main]
org.apache.cassandra.exceptions.ConfigurationException: Column family ID mismatch (found 8c5c36a0-7497-11ec-b2a0-f7a76f46da7d; expected 8bb2fe50-7497-11ec-9837-6fe6aca8f868)
        at org.apache.cassandra.config.CFMetaData.validateCompatibility(CFMetaData.java:1000) ~[apache-cassandra-3.11.11.jar:3.11.11]
        at org.apache.cassandra.config.CFMetaData.apply(CFMetaData.java:953) ~[apache-cassandra-3.11.11.jar:3.11.11]
        at org.apache.cassandra.config.Schema.updateTable(Schema.java:687) ~[apache-cassandra-3.11.11.jar:3.11.11]
        at org.apache.cassandra.schema.SchemaKeyspace.updateKeyspace(SchemaKeyspace.java:1478) ~[apache-cassandra-3.11.11.jar:3.11.11]
        at org.apache.cassandra.schema.SchemaKeyspace.mergeSchema(SchemaKeyspace.java:1434) ~[apache-cassandra-3.11.11.jar:3.11.11]
        at org.apache.cassandra.schema.SchemaKeyspace.mergeSchema(SchemaKeyspace.java:1403) ~[apache-cassandra-3.11.11.jar:3.11.11]
        at org.apache.cassandra.schema.SchemaKeyspace.mergeSchemaAndAnnounceVersion(SchemaKeyspace.java:1380) ~[apache-cassandra-3.11.11.jar:3.11.11]
        at org.apache.cassandra.db.DefinitionsUpdateVerbHandler$1.runMayThrow(DefinitionsUpdateVerbHandler.java:51) ~[apache-cassandra-3.11.11.jar:3.11.11]
        at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:28) ~[apache-cassandra-3.11.11.jar:3.11.11]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[na:1.8.0_312]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_312]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_312]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_312]
        at org.apache.cassandra.concurrent.NamedThreadFactory.lambda$threadLocalDeallocator$0(NamedThreadFactory.java:84) [apache-cassandra-3.11.11.jar:3.11.11]
        at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_312]
```

The same exception is also reported in Stargate's logs. 

When the CassandraDatacenter is configure to use racks, the Stargate controller will, by default, try to spread Stargate instances evenly across racks. It creates a separate Deployment per rack. The problem is that the Stargate instances were started too soon, one after the other. I confirmed this in the k8ssandra-operator logs:

`2022-01-13T17:36:33.945Z        INFO    controller-runtime.manager.controller.stargate  Stargate Deployment not found, creating a new one       {"reconciler group": "stargate.k8ssandra.io", "reconciler kind": "Stargate", "name": "s1", "namespace": "multi-stargate-fcb96g", "Deployment": {"namespace": "multi-stargate-fcb96g", "name": "test-dc1-rack1-stargate-deployment"}}`

`2022-01-13T17:36:34.005Z        INFO    controller-runtime.manager.controller.stargate  Stargate Deployment not found, creating a new one       {"reconciler group": "stargate.k8ssandra.io", "reconciler kind": "Stargate", "name": "s1", "namespace": "multi-stargate-fcb96g", "Deployment": {"namespace": "multi-stargate-fcb96g", "name": "test-dc1-rack2-stargate-deployment"}}`

`2022-01-13T17:36:34.085Z        INFO    controller-runtime.manager.controller.stargate  Stargate Deployment not found, creating a new one       {"reconciler group": "stargate.k8ssandra.io", "reconciler kind": "Stargate", "name": "s1", "namespace": "multi-stargate-fcb96g", "Deployment": {"namespace": "multi-stargate-fcb96g", "name": "test-dc1-rack3-stargate-deployment"}}`

Notice how close together in time `test-dc1-rack2-stargate-deployment` and `test-dc1-rack3-stargate-deployment` are created. We basically have a race condition with respect to initializing. 

This isn't a problem when Stargate is deployed as part of a K8ssandraCluster because the K8ssandraCluster controller initializes the Stargate `data_endpoint_auth` keyspace and the `data_endpoint_auth.token` before deploying Stargate; so, there is no race condition. While deploying Stargate on its own is probably a pretty low priority scenario, we definitely want to fix the flaky tests.

**Did you expect to see something different?**
If Stargate is deployed on its own and not part of a K8ssandraCluster, then the Stargate controller should wait to create the Deployment until the previously created Deployment becomes ready.

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-1170) by [Unito](https://www.unito.io)
┆friendlyId: K8SSAND-1170
┆priority: Medium

### PR DESCRIPTION
**Pull request by adutra**
**What this PR does**:

Yet Another Attempt (TM) to fix #297 :-) 

Implements @jsanda suggestion outlined here : https://github.com/k8ssandra/k8ssandra-operator/pull/304#issuecomment-1015073664

**Which issue(s) this PR fixes**:
Fixes #297

**Checklist**
- [ ] Changes manually tested
- [ ] Automated Tests added/updated
- [ ] Documentation added/updated
- [ ] CHANGELOG.md updated (not required for documentation PRs)
- [ ] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: controllers/k8ssandra/replication.go [hunk 1]
```diff
@@ -50,15 +50,20 @@ func (r *K8ssandraClusterReconciler) checkSystemReplication(ctx context.Context,
...
+) result.ReconcileResult {
+	managementApiFacade, err := r.ManagementApi.NewManagementApiFacade(ctx, dcs[0], remoteClient, logger)
 	if err != nil {
 		logger.Error(err, "Failed to create ManagementApiFacade")
 		return result.Error(err)
 	}
 
 	keyspaces := []string{"system_traces", "system_distributed", "system_auth"}
-	replication := cassandra.ComputeReplication(3, kc.Spec.Cassandra.Datacenters...)
+	replication := cassandra.ComputeReplication(3, dcs...)
```
Comment (adutra) [inline controllers/k8ssandra/replication.go:66]: So, this changes the semantics slightly for multi-dc setups.

Previously: the keyspaces would be created with the desired replication for all dcs, including those not yet created.

Now: the keyspaces are created with the desired replication for the first dc, then updated to include replication for the second dc, and so on.

@jsanda do you think this is acceptable?

Comment (jsanda) [reply controllers/k8ssandra/replication.go:66]: This won't work. I somewhat explain it in https://github.com/k8ssandra/k8ssandra-operator/pull/262#discussion_r784222098. In short we would need to run repair on the system_auth keyspace on the second dc. Without doing that, the second dc won't be able to properly initialize.

Comment (adutra) [reply controllers/k8ssandra/replication.go:66]: Sounds good, I will fix that.

Comment (adutra) [reply controllers/k8ssandra/replication.go]: It's fixed now. Out of curiosity, why do we need to call this methods for every dc, wouldn't it be enough to call it for the first dc only?

Comment (jsanda) [reply controllers/k8ssandra/replication.go]: Great questions. There are a couple reasons it is called for every DC. First, we need to handle scale up within the DC. If we start with a single node cluster, system_auth is configured with RF = 1. If we then scale up to 3 nodes, it needs to be updated to RF = 3. Repair is needed as well, but that'll have to come post-1.0. The second reason has to do with adding a DC to an existing cluster. The are more, related changes around this in my add-dc branch. If we add dc2 to an existing cluster, again the RF needs to be updated.

File: controllers/k8ssandra/replication.go [hunk 2]
```diff
@@ -58,7 +58,7 @@ func (r *K8ssandraClusterReconciler) updateReplicationOfSystemKeyspaces(ctx cont
 	}
 
 	keyspaces := []string{"system_traces", "system_distributed", "system_auth"}
-	replication := cassandra.ComputeReplication(3, kc.Spec.Cassandra.Datacenters...)
+	replication := cassandra.ComputeReplicationFromDcTemplates(3, kc.Spec.Cassandra.Datacenters...)
```
Comment (jsanda) [inline controllers/k8ssandra/replication.go:61]: FYI, I change this is #262. See [here](https://github.com/k8ssandra/k8ssandra-operator/blob/0d3f33ea3fbf29fe48f0a7d8d49ddab38e62eac9/pkg/cassandra/util.go#L36) for details.

File: controllers/stargate/stargate_controller.go [hunk 3]
```diff
@@ -144,6 +145,17 @@ func (r *StargateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 			"Stargate", req.NamespacedName)
 	}
 
+	if !labels.IsOwnedByK8ssandraController(stargate) {
```
Comment (jsanda) [inline controllers/stargate/stargate_controller.go:148]: 👌

**Review by jsanda — APPROVED** [review]
This is probably going to cause some conflicts for me in #262; however, this is definitely my preferred solution. Thanks for working through each of them!

**Pull request merged**